src/utils/dockerDirManager.ts

```diff
@@ -0,0 +1,302 @@
+   */
+  protected async getCurrentDockerSocket(currentContext?: string): Promise<string> {
+    const defaultSocket = `unix://${ this.defaultDockerSockPath }`;
+
+    if (!currentContext) {
+      return defaultSocket;
+    }
+
+    for (const dir of await fs.promises.readdir(this.dockerContextDirPath)) {
+      const contextPath = path.join(this.dockerContextDirPath, dir, 'meta.json');
+
+      try {
+        const data = yaml.parse(await fs.promises.readFile(contextPath, 'utf-8'));
+
+        if (data.Name === currentContext) {
+          return data.Endpoints?.docker?.Host as string ?? defaultSocket;
+        }
+      } catch (ex) {
+        console.log(`Failed to read context ${ dir }, skipping: ${ ex }`);
+      }
+    }
+
+    // If we reach here, the current context is invalid.
+    return defaultSocket;
+  }
+
+  /**
+   * Given some information about state external to this method, returns the
+   * name of the context that should be used. Follows these rules, in order of preference:
+   * 1. If we have control of the default socket (`/var/run/docker.sock`), we should set the
+   *    context to it (which is actually un-setting the `currentContext` key).
+   *    This should have the widest compatibility.
+   * 2. Otherwise, check the current context and don't change anything if any of the following
+   *    is true:
+   *    - The current context uses a valid unix socket - the user is probably using it.
+   *    - The current context uses a non-unix socket (e.g. tcp) - we can't check if it's valid.
+   * 3. The current context is invalid - set the current context to our (rancher-desktop) context.
+   * @param weOwnDefaultSocket Whether Rancher Desktop has control over the default socket.
+   * @param currentContext The current context.
+   */
+  async getDesiredDockerContext(weOwnDefaultSocket: boolean, currentContext: string | undefined): Promise<string | undefined> {
+    if (weOwnDefaultSocket) {
+      return undefined;
+    }
+
+    if (!currentContext && !weOwnDefaultSocket) {
+      return this.contextName;
+    }
+
+    if (currentContext === this.contextName) {
+      return this.contextName;
+    }
+
+    const currentSocketUri = await this.getCurrentDockerSocket(currentContext);
+
+    if (!currentSocketUri.startsWith('unix://')) {
+      // Using a non-unix socket (e.g. TCP); assume it's working fine.
+      return currentContext;
+    }
+
+    const currentSocketPath = currentSocketUri.replace(/^unix:\/\//, '');
+
+    try {
+      if ((await fs.promises.stat(currentSocketPath)).isSocket()) {
+        return currentContext;
+      }
+      console.log(`Invalid existing context "${ currentContext }": ${ currentSocketUri } is not a socket; overriding context.`);
+    } catch (ex) {
+      console.log(`Could not read existing docker socket ${ currentSocketUri }, overriding context "${ currentContext }": ${ ex }`);
+    }
+
+    return this.contextName;
+  }
+
+  /**
+   * Determines whether the passed cred helper is working.
```

Comment:
```suggestion
   * Determines whether the passed credential helper is working.
```
(Should avoid abbreviations on first use in a section.)